Sources/GRPC/GRPCKeepaliveHandlers.swift

```diff
@@ -0,0 +1,281 @@
+    minimumReceivedPingIntervalWithoutData: TimeAmount? = nil,
+    maximumPingStrikes: UInt? = nil
+  ) {
+    self.pingCode = pingCode
+    self.time = time
+    self.timeout = timeout
+    self.permitWithoutCalls = permitWithoutCalls
+    self.maximumPingsWithoutData = maximumPingsWithoutData
+    self.minimumSentPingIntervalWithoutData = minimumSentPingIntervalWithoutData
+    self.minimumReceivedPingIntervalWithoutData = minimumReceivedPingIntervalWithoutData
+    self.maximumPingStrikes = maximumPingStrikes
+  }
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    if event is NIOHTTP2StreamCreatedEvent && self.activeStreams == 0 {
+      self.activeStreams += 1
+      self.schedulePing(context: context)
+    } else if event is StreamClosedEvent {
+      self.activeStreams -= 1
+    }
+  }
+
+  func read(pingData: HTTP2PingData, ack: Bool, context: ChannelHandlerContext) {
+    if ack && pingData.integer == self.pingCode {
+      // Received pong
+      self.acknowledgePong()
+    } else if !ack && self.activeStreams == 0 && !self.permitWithoutCalls {
+      // No calls in flight, send a GOAWAY / ENHANCE_YOUR_CALM to client and close connection
+      self.sendGoAway(context: context)
+      self.channelHandler?.close(context: context)
+    } else if !ack {
+      // Received ping, send pong
+      self.lastReceivedPingDate = .now()
+      self.sendPing(code: pingData.integer, ack: true, context: context)
+    }
+  }
+
+  private func schedulePing(context: ChannelHandlerContext) {
+    self.scheduledPing = context.eventLoop.scheduleRepeatedTask(initialDelay: self.time, delay: self.time) { _ in
+      if !self.shouldBlockPing() {
+        self.sendPing(code: self.pingCode, ack: false, context: context)
+        self.scheduleClose(context: context)
+      }
+    }
+  }
+
+  private func scheduleClose(context: ChannelHandlerContext) {
+    self.scheduledClose = context.eventLoop.scheduleTask(in: self.timeout) {
+      self.channelHandler?.close(context: context)
+    }
+  }
+
+  private func sendPing(code: UInt64, ack: Bool, context: ChannelHandlerContext) {
+    if self.activeStreams == 0 {
+      self.sentPingsWithoutData += 1
+    }
+
+    let payload = HTTP2Frame.FramePayload.ping(HTTP2PingData(withInteger: code), ack: ack)
+    self.channelHandler?.send(payload: payload, context: context)
+    lastSentPingDate = .now()
+  }
+
+  private func sendGoAway(context: ChannelHandlerContext) {
+    let payload = HTTP2Frame.FramePayload.goAway(lastStreamID: .rootStream, errorCode: .enhanceYourCalm, opaqueData: nil)
+    self.channelHandler?.send(payload: payload, context: context)
+  }
+
+  private func acknowledgePong() {
+    self.scheduledClose?.cancel()
+    self.scheduledClose = nil
+  }
+
+  private func shouldBlockPing() -> Bool {
+    // There is no active call on the transport and pings should not be sent
+    if self.activeStreams == 0 && !self.permitWithoutCalls { return true }
```

Comment:
This might be slightly easier to follow if we `guard self.activeStreams == 0 else { return false }` and then deal with `permitWithoutCalls`.

nit: generally we don't have one-liners like this, please move the return to a new line.